File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第8章找素数.py
#找素数 lanqiaoOJ 1558

from math import *
N = 1000005
vis = [True]*N           # 标记[2,sqrt(b)]是否为素数
prime = [0]*N            # 存[a,b]的素数
seg_prime = [True]*(N)   # 标记[a,b]是否为素数
def seg_sieve(a,b):
    for i in range(2,int(sqrt(b))+1):
        if vis[i]:    # 是素数
            for j in range(i*i,int(sqrt(b)),i):        vis[j]=False
            for j in range(max(2,(a+i-1)//i)*i,b+1,i): seg_prime[j-a]=False   
    num = 0
    for i in range(0,b-a+1):
        if seg_prime[i]:
            prime[num] = i+a
            num += 1
    print(num)    
a,b = map(int, input().split())
seg_sieve(a, b)

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第8章找素数.java
//找素数   lanqiaoOJ 1558 

import java.util.*;

public class Main {
    public static final int MAX=1000005;
    public static boolean isprime[]=new boolean[MAX];
    public static boolean primelist[]=new boolean[MAX];
    
    public static int seg_sieve(long a,long b) {
        for(long i=0;i*i<=b;i++) {//此处要用long定义i，否则 i*i会越界
            isprime[(int) i]=true;            
        }
        for(int i=0;i<=b-a;i++) {
            primelist[i]=true;
        }
        for(long i=2;i*i<=b;i++) {
            if(isprime[(int) i]) {                  //i是素数
                for(long j=i*i;j*j<=b;j+=i)//筛选[2,sqrt(b)]内的素数
                    isprime[(int) j] = false;
                 for(long j=Math.max(2,(a+i-1)/i)*i;j<=b;j+=i)//筛选[a,b]内的素数
                     primelist[(int) (j-a)] = false;
            }
        }
        int k=0;
        for(int i=0;i<=b-a;++i) {
            if(primelist[i]) k++;
        }
        return k;
    }
    public static void main(String[] args) {
        Scanner sc=new Scanner(System.in);
        int a = sc.nextInt();
        int b = sc.nextInt();
        int k = seg_sieve(a,b);        
        
        System.out.println(k);
    }
}